### PR TITLE
E-822-migrate-rewardable-items

### PR DESCRIPTION
**Pull request by pixelfact**

### CODE REVIEW

File: src/pages/ProjectDetails/Rewards/RewardForm/WorkItemSidePanel/WorkItems/WorkItems.tsx [hunk 1]
```diff
@@ -13,6 +13,14 @@ import {
 import View from "./View";
 import { useIgnoredContributions } from "./useIgnoredContributions";
 import { useApolloClient } from "@apollo/client";
+import useRewardableItemsQueryParams from "../hooks/useRewardableItemsQueryParams";
+import ProjectApi from "src/api/Project";
+
+export enum ContributionType {
```
Comment (haydencleary) [inline src/pages/ProjectDetails/Rewards/RewardForm/WorkItemSidePanel/WorkItems/WorkItems.tsx:19]: This should already exist in the `types.ts` file

File: src/pages/ProjectDetails/Rewards/RewardForm/utils.ts [hunk 2]
```diff
@@ -1,17 +1,17 @@
...
-): ContributionFragment[] => {
+  contributions: RewardableItem[]
+): RewardableItem[] => {
   return filter(contributions, {
-    status: "complete",
+    status: "COMPLETED",
     type,
     ignored: false,
-  }) as ContributionFragment[];
+  }) as RewardableItem[];
```
Comment (haydencleary) [inline src/pages/ProjectDetails/Rewards/RewardForm/utils.ts:14]: Are a return type _and_ a cast type necessary ?

File: src/pages/ProjectDetails/Rewards/RewardForm/WorkItemSidePanel/hooks/useRewardableItemsQueryParams.tsx [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  const [includeIgnoredItems, setIncludeIgnoredItems] = useState<boolean>(props.includeIgnoredItems || false);
+
+  const queryParams: QueryParams = useMemo(() => {
+    const params: Record<string, string> = {};
+
+    if (type) params["type"] = type;
+    if (search) params["search"] = search;
+    if (includeIgnoredItems || includeIgnoredItems === false)
+      params["include_ignored_items"] = includeIgnoredItems.toString();
+    if (props.githubUserId !== undefined) params["githubUserId"] = props.githubUserId.toString();
```
Comment (haydencleary) [inline src/pages/ProjectDetails/Rewards/RewardForm/WorkItemSidePanel/hooks/useRewardableItemsQueryParams.tsx:22]:
```suggestion
    if (props.githubUserId) params["githubUserId"] = props.githubUserId.toString();
```

File: src/pages/ProjectDetails/Rewards/RewardForm/WorkItemSidePanel/hooks/useRewardableItemsQueryParams.tsx [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  const [type, setType] = useState<string>(props.type || "");
+  const [search, setSearch] = useState<string>("");
+  const [includeIgnoredItems, setIncludeIgnoredItems] = useState<boolean>(props.includeIgnoredItems || false);
+
+  const queryParams: QueryParams = useMemo(() => {
+    const params: Record<string, string> = {};
+
+    if (type) params["type"] = type;
+    if (search) params["search"] = search;
+    if (includeIgnoredItems || includeIgnoredItems === false)
```
Comment (haydencleary) [inline src/pages/ProjectDetails/Rewards/RewardForm/WorkItemSidePanel/hooks/useRewardableItemsQueryParams.tsx:20]: Doesn't this condition pass every time ?

File: src/pages/ProjectDetails/Rewards/RewardForm/WorkItemSidePanel/WorkItems/WorkItems.tsx (hunk 1 above)
Comment (pixelfact) [reply src/pages/ProjectDetails/Rewards/RewardForm/WorkItemSidePanel/WorkItems/WorkItems.tsx]: yes right didn't see it at this time, delete it in next commits

File: src/pages/ProjectDetails/Rewards/RewardForm/utils.ts (hunk 2 above)
Comment (pixelfact) [reply src/pages/ProjectDetails/Rewards/RewardForm/utils.ts:14]: nop !

File: src/components/GithubCard/GithubIssue/GithubIssue.tsx [hunk 5]
```diff
@@ -18,15 +19,14 @@ export enum Action {
...
-function getIssueStatusDate(issue: GithubIssueFragment) {
+function getIssueStatusDate(issue: Partial<GithubIssueFragment & RewardableItem>) {
   switch (issue.status) {
     case GithubIssueStatus.Cancelled:
     case GithubIssueStatus.Completed:
     case ContributionStatus.Completed:
     case ContributionStatus.Cancelled:
       return new Date(issue.closedAt);
     case GithubIssueStatus.Open:
-    case ContributionStatus.InProgress:
```
Comment (pixelfact) [inline src/components/GithubCard/GithubIssue/GithubIssue.tsx:29]: @haydencleary this status doesn't exist anymore in the code gen, so i had to delete it. Im i wrong ?!

File: src/components/GithubCard/GithubPullRequest/GithubPullRequest.tsx [hunk 6]
```diff
@@ -34,7 +37,6 @@ function getPullRequestStatusDate(pullRequest: GithubPullRequestWithCommitsFragm
     case ContributionStatus.Completed:
       return new Date(pullRequest.mergedAt);
     case GithubPullRequestStatus.Open:
-    case ContributionStatus.InProgress:
```
Comment (pixelfact) [inline src/components/GithubCard/GithubPullRequest/GithubPullRequest.tsx:37]: @haydencleary same : this status doesn't exist anymore in the code gen, so i had to delete it. Im i wrong ?!

File: src/pages/ProjectDetails/Rewards/RewardForm/index.tsx [hunk 7]
```diff
@@ -76,15 +78,25 @@ const RewardForm: React.FC = () => {
...
+    isError,
+  } = ProjectApi.queries.useRewardableItemsInfiniteList({
+    // WE need to fetch all the contributions to be able to AUTO-ADD them all in one click
+    // It's the reason that we set pageSize to 1000 assuming that there will never be more than 1000 contributions
+    // and in the case there are more than 1000 contributions, we assume this limitation for performance reasons
+    params: { projectId, queryParams, pageSize: 1000 },
+    options: { enabled: !!contributor?.githubUserId },
+  });
+
+  const contributions = contributionItems?.pages.flatMap(({ rewardableItems }) => rewardableItems) ?? [];
```
Comment (pixelfact) [inline src/pages/ProjectDetails/Rewards/RewardForm/index.tsx:98]: not tested, waiting for backend to provide a dedicated endpoint to be used to auto add and ignore contributions

File: src/pages/ProjectDetails/Rewards/RewardForm/WorkItemSidePanel/ErrorState/index.tsx
Comment (haydencleary) [inline src/pages/ProjectDetails/Rewards/RewardForm/WorkItemSidePanel/ErrorState/index.tsx:1]: Maybe relocate if its a generic component

File: src/pages/ProjectDetails/Rewards/RewardForm/WorkItemSidePanel/OtherWorkForm/index.tsx [hunk 8]
```diff
@@ -64,42 +63,57 @@ export default function OtherWorkForm({ projectId, contributorHandle, addWorkIte
...
   const onSubmit: FormEventHandler<HTMLFormElement> = e => {
     e.preventDefault();
-    handleSubmit(() => createIssue())(e);
+    handleSubmit(() =>
+      createOtherWork({
+        githubRepoId: selectedRepo?.id || 0,
+        title: title || defaultTitle,
+        description,
+      })
+    )(e);
```
Comment (haydencleary) [inline src/pages/ProjectDetails/Rewards/RewardForm/WorkItemSidePanel/OtherWorkForm/index.tsx:116]: Don't think you need the event param here

File: src/pages/ProjectDetails/Rewards/RewardForm/utils.ts [hunk 9]
```diff
@@ -1,17 +1,17 @@
...
 
 export const filterUnpaidContributionsByType = (
   type: GithubContributionType,
-  contributions: ContributionFragment[]
-): ContributionFragment[] => {
+  contributions: RewardableItem[]
+): RewardableItem[] => {
   return filter(contributions, {
-    status: "complete",
+    status: "COMPLETED",
```
Comment (haydencleary) [inline src/pages/ProjectDetails/Rewards/RewardForm/utils.ts:11]: Use enum

File: src/translations/en.json [hunk 10]
```diff
@@ -459,6 +459,7 @@
             "noIssues": "This repository does not allow creating issues, please change this setting on Github to add other work."
           },
           "success": "Issue created and contribution added",
+          "error": "The Issue could not be created!",
```
Comment (haydencleary) [inline src/translations/en.json:462]:
```suggestion
          "error": "The issue could not be created!",
```

File: src/utils/api.ts [hunk 11]
```diff
@@ -50,9 +51,11 @@ export function formatRewardItemToGithubIssue(item: components["schemas"]["Rewar
   } as GithubIssueFragment;
 }
 
-export function formatRewardItemToGithubCodeReview(item: components["schemas"]["RewardItemResponse"]) {
+export function formatRewardItemToGithubCodeReview(
+  item: Partial<components["schemas"]["RewardItemResponse"] & RewardableItem>
```
Comment (haydencleary) [inline src/utils/api.ts:55]: Use the exported type `RewardableItem`

File: src/api/Project/queries.ts [hunk 12]
```diff
@@ -51,4 +55,101 @@ const useInfiniteList = (
...
+
+  const debounceSearch = useCallback(
+    debounce(newSearch => {
+      setDebouncedSearch(newSearch);
+    }, 300),
+    []
+  );
+
+  useEffect(() => {
+    if (search || search === "") {
```
[the remark this answers is not in the document]
Comment (haydencleary) [reply src/api/Project/queries.ts:103]: Same modification in a different file, I'll let you look

File: src/utils/api.ts (hunk 11 above)
Comment (pixelfact) [reply src/utils/api.ts:55]: not the same type !

**Review by haydencleary — APPROVED** [review]

**Pull request merged**